sql/updates/latest-dev.sql

```diff
@@ -0,0 +1,47 @@
+-- Dimension type used in create_hypertable, add_dimension, etc. It is
+-- deliberately an opaque type.
+CREATE TYPE @extschema@.dimension_info;
+
+CREATE FUNCTION @extschema@.dimension_info_in(cstring) RETURNS dimension_info
+AS '@MODULE_PATHNAME@', 'ts_dimension_info_in' LANGUAGE C STRICT IMMUTABLE;
+
+CREATE FUNCTION @extschema@.dimension_info_out(dimension_info) RETURNS cstring
+AS '@MODULE_PATHNAME@', 'ts_dimension_info_in' LANGUAGE C STRICT IMMUTABLE;
+
+CREATE TYPE @extschema@.dimension_info (
+    INPUT = dimension_info_in,
+    OUTPUT = dimension_info_out,
+    INTERNALLENGTH = VARIABLE
+);
+
+CREATE OR REPLACE FUNCTION @extschema@.ByHash(column_name NAME, number_partitions INTEGER,
+                                            partition_func regproc = NULL)
+RETURNS dimension_info LANGUAGE C AS '@MODULE_PATHNAME@', 'ts_hash_dimension';
+
+CREATE OR REPLACE FUNCTION @extschema@.ByRange(column_name NAME,
+                                             partition_interval ANYELEMENT = NULL::bigint,
+                                             partition_func regproc = NULL)
+RETURNS dimension_info LANGUAGE C AS '@MODULE_PATHNAME@', 'ts_range_dimension';
+
+-- API changes related to hypertable generalization
+CREATE OR REPLACE FUNCTION @extschema@.create_hypertable(
```

Comment:
I do not think we should break the existing  create_hypertable API.
We can extend the API, but not break it. Especially by introducing a complicated type like dimension_info.

I like the fact that this is not experimental and part of the regular APIs. But we should no break an existing API without any deprecation.

Reply:
> I do not think we should break the existing create_hypertable API. We can extend the API, but not break it. Especially by introducing a complicated type like dimension_info.
>
> I like the fact that this is not experimental and part of the regular APIs. But we should no break an existing API without any deprecation.

We are not breaking the interface, we are extending it. All existing tests pass without changes. The old function calls work just as before, and the new type is there to make sure that overload resolution pick the right function.

Reply:
Since we do not want to place ourselves in a position where we have to rename the functions some time in the future, I think this approach is better compared to putting functions in the experimental schema.

This function should be marked as experimental in the documentation, but we might want to print a warning as well that the interface might change (basically saying that this new interface is experimental).

Reply:
Thank you for clarifying.